### ISSUE TITLE
The same buffer potentially used in 2 SSE channels

### ISSUE THREAD

**Issue by martintrojer** [pedestal/pedestal#369]
If a buffer is passed into `start-stream` (via `start-event-stream`) 2 channels is created with the same buffer instance. I don't know if that's intentional or what core.async does in that situation.

Its used here https://github.com/pedestal/pedestal/blob/master/service/src/io/pedestal/http/sse.clj#L140 and here https://github.com/pedestal/pedestal/blob/master/service/src/io/pedestal/http/sse.clj#L148

**Comment by ohpauleez** [comment]
While the arg is called "buffer-or-n", it was only intended to be "n", which would be ok in this case.  If you passed a buffer in, core.async would crash.

If you're trying to us a buffer here, each channel really needs its own buffer (and we'd need to patch the code).

**Comment by martintrojer** [comment]
I'd love to be able to pass in sliding buffers here tho.

Indeed, I get stack traces like below.
```
Exception in thread "async-dispatch-41" java.util.NoSuchElementException
	at java.util.LinkedList.removeLast(LinkedList.java:283)
	at clojure.core.async.impl.buffers.SlidingBuffer.remove_BANG_(buffers.clj:57)
	at clojure.core.async.impl.channels.ManyToManyChannel.take_BANG_(channels.clj:173)
	at clojure.core.async.impl.ioc_macros$take_BANG_.invoke(ioc_macros.clj:950)
	at io.pedestal.http.sse$start_stream$fn__28815$state_machine__7037__auto____28816$fn__28818.invoke(sse.clj:126)
	at io.pedestal.http.sse$start_stream$fn__28815$state_machine__7037__auto____28816.invoke(sse.clj:124)
	at clojure.core.async.impl.ioc_macros$run_state_machine.invoke(ioc_macros.clj:940)
	at clojure.core.async.impl.ioc_macros$run_state_machine_wrapped.invoke(ioc_macros.clj:944)
	at io.pedestal.http.sse$start_stream$fn__28815.invoke(sse.clj:124)
```

**Comment by ohpauleez** [comment]
I'll look at this right now - thanks for the report!

**Comment by ohpauleez** [comment]
Fixed on master.  Since no one could have working code that used the "buffer", I made the arg "bufferfn-or-n".  You can now put a zero-arity function in the buffer position that returns a new core.async buffer when called.  Fixed buffers still work as before.

**Comment by martintrojer** [comment]
:+1: 

How far away is next release? :)

**Comment by ohpauleez** [comment]
We were trying to wait until three larger changes landed, allowing us to skip 0.4.1 and release 0.5.0.  Those changes are going to require much more work - I'll look into cutting an 0.4.1 release with the current changes (lots of nice, small improvements on master).

**Comment by martintrojer** [comment]
I'd be eternally grateful :)
